In [1]:
!pip install apyori

  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5975 sha256=dce5da53c04a9d0acc597a2da03f083707732f2af790f24381e5a2c75fda6f08
  Stored in directory: c:\users\vaibh\appdata\local\pip\cache\wheels\84\45\a4\8ade6576f75410d8162c6da1de0aa9df56c16c711acab5a813
Successfully built apyori


Objective of Market Basket
1. Cross Selling: It is a stratergy where the seller encourages the customer to spend more money by recommending related products that complement what is being bought already by the consumer. This stratergy would encourage the customer to spend more than he/she had actually thought he/she would.

2. Product Placement: When you go to supermarket you may see that the milk items are kept together. Moreover you may see that as you move forward you find the bread making items such as flour, butter, eggs etc kept just after the milk products. This placement of the items on the supermarket shelf follows planogram. A planogram is defined as a “diagram or model that indicates the placement of retail products on shelves in order to maximize sales”.

In [2]:
import numpy as np # linear algebra
import pandas as pd # data processing

from apyori import apriori

In [4]:
df  = pd.read_csv('Market_Basket_Optimisation.csv',header=None)
df.head() 

0          1           2                 3             4   \
0         shrimp    almonds     avocado    vegetables mix  green grapes   
1        burgers  meatballs        eggs               NaN           NaN   
2        chutney        NaN         NaN               NaN           NaN   
3         turkey    avocado         NaN               NaN           NaN   
4  mineral water       milk  energy bar  whole wheat rice     green tea   

                 5     6               7             8             9   \
0  whole weat flour  yams  cottage cheese  energy drink  tomato juice   
1               NaN   NaN             NaN           NaN           NaN   
2               NaN   NaN             NaN           NaN           NaN   
3               NaN   NaN             NaN           NaN           NaN   
4               NaN   NaN             NaN           NaN           NaN   

               10         11     12     13             14      15  \
0  low fat yogurt  green tea  honey  salad  mineral water  salmon   
1             NaN        NaN    NaN    NaN            NaN     NaN   
2             NaN        NaN    NaN    NaN            NaN     NaN   
3             NaN        NaN    NaN    NaN            NaN     NaN   
4             NaN        NaN    NaN    NaN            NaN     NaN   

                  16               17       18         19  
0  antioxydant juice  frozen smoothie  spinach  olive oil  
1                NaN              NaN      NaN        NaN  
2                NaN              NaN      NaN        NaN  
3                NaN              NaN      NaN        NaN  
4                NaN              NaN      NaN        NaN

In [5]:
df.shape

(7501, 20)

In [6]:
df.isnull().sum()

0        0
1     1754
2     3112
3     4156
4     4972
5     5637
6     6132
7     6520
8     6847
9     7106
10    7245
11    7347
12    7414
13    7454
14    7476
15    7493
16    7497
17    7497
18    7498
19    7500
dtype: int64

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7501 entries, 0 to 7500
Data columns (total 20 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       7501 non-null   object
 1   1       5747 non-null   object
 2   2       4389 non-null   object
 3   3       3345 non-null   object
 4   4       2529 non-null   object
 5   5       1864 non-null   object
 6   6       1369 non-null   object
 7   7       981 non-null    object
 8   8       654 non-null    object
 9   9       395 non-null    object
 10  10      256 non-null    object
 11  11      154 non-null    object
 12  12      87 non-null     object
 13  13      47 non-null     object
 14  14      25 non-null     object
 15  15      8 non-null      object
 16  16      4 non-null      object
 17  17      4 non-null      object
 18  18      3 non-null      object
 19  19      1 non-null      object
dtypes: object(20)
memory usage: 1.1+ MB


In [9]:
df.head()

0          1           2                 3             4   \
0         shrimp    almonds     avocado    vegetables mix  green grapes   
1        burgers  meatballs        eggs               NaN           NaN   
2        chutney        NaN         NaN               NaN           NaN   
3         turkey    avocado         NaN               NaN           NaN   
4  mineral water       milk  energy bar  whole wheat rice     green tea   

                 5     6               7             8             9   \
0  whole weat flour  yams  cottage cheese  energy drink  tomato juice   
1               NaN   NaN             NaN           NaN           NaN   
2               NaN   NaN             NaN           NaN           NaN   
3               NaN   NaN             NaN           NaN           NaN   
4               NaN   NaN             NaN           NaN           NaN   

               10         11     12     13             14      15  \
0  low fat yogurt  green tea  honey  salad  mineral water  salmon   
1             NaN        NaN    NaN    NaN            NaN     NaN   
2             NaN        NaN    NaN    NaN            NaN     NaN   
3             NaN        NaN    NaN    NaN            NaN     NaN   
4             NaN        NaN    NaN    NaN            NaN     NaN   

                  16               17       18         19  
0  antioxydant juice  frozen smoothie  spinach  olive oil  
1                NaN              NaN      NaN        NaN  
2                NaN              NaN      NaN        NaN  
3                NaN              NaN      NaN        NaN  
4                NaN              NaN      NaN        NaN

In [10]:
#converting dataframe into list of lists
dataframe_list=[]

# values of range is according to the number of rows and columns
for i in range(1,7501):
    dataframe_list.append([str(df.values[i,j]) for j in range(0,20)])

In [11]:
#applying apriori algorithm
association_rules = apriori(dataframe_list, min_support=0.0045, min_confidence=0.2,
                            min_lift=3, min_length=2)
association_results = list(association_rules)

In [12]:
association_results[0]

RelationRecord(items=frozenset({'light cream', 'chicken'}), support=0.004533333333333334, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.2905982905982906, lift=4.843304843304844)])

Inferences from the above
Support: 0.0045.
The support is calculated = (chicken + light cream)/(total transactions)

Confidence: 0.2905
Confidence = (chicken + light cream)/(light cream)
The confidence level for the rule is 0.2905, which shows that out of all the transactions that contain light cream, 29.05% percent contain chicken too.

Lift: 4.84
Lift = ((chicken + light cream)/light cream)/((chicken)/total)
The lift of 4.8433 tells us that chicken is 4.8433 times more likely to be bought by the customers who buy light cream compared to the default likelihood sale of chicken.

In [13]:
# Creating a DataFrame

In [14]:
Rule = []
Support = []
Confidence = []
Lift = []
for item in association_results:
    pair = item[0]
    items = [x for x in pair]
    Rule.append(items)
    Support.append(str(item[1]))
    Confidence.append(str(item[2][0][2]))
    Lift.append(str(item[2][0][3]))

In [15]:
dict = {'Rule': Rule, 'Support': Support, 'Confidence': Confidence, "Lift": Lift} 
d = pd.DataFrame(dict)
d

Rule               Support  \
0                              [light cream, chicken]  0.004533333333333334   
1                    [escalope, mushroom cream sauce]  0.005733333333333333   
2                                   [pasta, escalope]  0.005866666666666667   
3                        [herb & pepper, ground beef]                 0.016   
4                         [tomato sauce, ground beef]  0.005333333333333333   
5                      [whole wheat pasta, olive oil]                 0.008   
6                                     [pasta, shrimp]  0.005066666666666666   
7                         [nan, light cream, chicken]  0.004533333333333334   
8              [shrimp, chocolate, frozen vegetables]  0.005333333333333333   
9               [cooking oil, spaghetti, ground beef]                0.0048   
10              [escalope, nan, mushroom cream sauce]  0.005733333333333333   
11                             [pasta, escalope, nan]  0.005866666666666667   
12        [spaghetti, frozen vegetables, ground beef]  0.008666666666666666   
13               [milk, olive oil, frozen vegetables]                0.0048   
14         [mineral water, shrimp, frozen vegetables]                0.0072   
15          [spaghetti, olive oil, frozen vegetables]  0.005733333333333333   
16             [shrimp, spaghetti, frozen vegetables]                 0.006   
17           [tomatoes, spaghetti, frozen vegetables]  0.006666666666666667   
18            [grated cheese, spaghetti, ground beef]  0.005333333333333333   
19        [herb & pepper, mineral water, ground beef]  0.006666666666666667   
20                  [herb & pepper, nan, ground beef]                 0.016   
21            [herb & pepper, spaghetti, ground beef]                0.0064   
22                     [milk, olive oil, ground beef]  0.004933333333333333   
23                   [tomato sauce, nan, ground beef]  0.005333333333333333   
24                   [shrimp, spaghetti, ground beef]                 0.006   
25                       [milk, spaghetti, olive oil]                0.0072   
26                   [soup, olive oil, mineral water]                0.0052   
27                [nan, olive oil, whole wheat pasta]                 0.008   
28                               [pasta, shrimp, nan]  0.005066666666666666   
29                   [spaghetti, olive oil, pancakes]  0.005066666666666666   
30        [shrimp, nan, chocolate, frozen vegetables]  0.005333333333333333   
31         [nan, cooking oil, spaghetti, ground beef]                0.0048   
32   [nan, spaghetti, frozen vegetables, ground beef]  0.008666666666666666   
33  [milk, spaghetti, frozen vegetables, mineral w...  0.004533333333333334   
34          [milk, nan, olive oil, frozen vegetables]                0.0048   
35    [mineral water, shrimp, nan, frozen vegetables]                0.0072   
36     [spaghetti, nan, olive oil, frozen vegetables]  0.005733333333333333   
37        [shrimp, nan, spaghetti, frozen vegetables]                 0.006   
38      [tomatoes, nan, spaghetti, frozen vegetables]  0.006666666666666667   
39       [nan, grated cheese, spaghetti, ground beef]  0.005333333333333333   
40   [herb & pepper, nan, mineral water, ground beef]  0.006666666666666667   
41       [herb & pepper, nan, spaghetti, ground beef]                0.0064   
42                [milk, nan, olive oil, ground beef]  0.004933333333333333   
43              [shrimp, nan, spaghetti, ground beef]                 0.006   
44                  [milk, spaghetti, nan, olive oil]                0.0072   
45              [nan, olive oil, mineral water, soup]                0.0052   
46              [spaghetti, nan, olive oil, pancakes]  0.005066666666666666   
47  [nan, frozen vegetables, milk, spaghetti, mine...  0.004533333333333334   

             Confidence                Lift  
0    0.2905982905982906   4.843304843304844  
1   0.30069930069930073  3.7903273197390845  
2   0.37288135593220345   4.70018515880

Conclusions
1. In this tutorial we went through the market basket analysis using Apriori Algorithm.
2. Study the key metrics used in algorithm
3. Check the assosiation rules and infered the values we got
4. Summariase the result as a dataframe